### PR TITLE
Liquidation debt

### PR DESCRIPTION
**Pull request by ith-harvey**
Functionality:
* Added `t0DebtInAuction` accumulator to track total debt in auction at a given time (accrues interest at pool rate)
* Added `_isActive()` back to `auctions.sol` as to perform correct `t0DebtInAuction` accum / decrem
* Added `_isAuctionDebtLocked()` to `pool.sol` which performs a check when LPB is removed from the following methods:
  * `removeQuoteToken()`
  * `removeAllQuoteTOken()`
  * `moveQuoteToken()`

Testing:
* Added `AuctionState` struct to `dsTestPlus.sol` to avoid stack too deep err

View:
* Added `debtInfo()` to `pool.sol` so `debtInAuction`can be checked in `_assertAuction`
* Removed `accruedDebt()` from `pool.sol` instead included in `debtInfo()`
* Added `_isAuctionDebtLocked` to `lenderInfo()`

Gas increases : ( :
`develop`
`liquidation-debt`

`moveQuoteToken                             ┆ 288488          ┆ 288488 ┆ 288488 ┆ 288488  ┆ 1  `
`moveQuoteToken                             ┆ 274831          ┆ 274831 ┆ 274831 ┆ 274831  ┆ 1   `
 
`pledgeCollateral                           ┆ 62657           ┆ 64767  ┆ 64836  ┆ 508950  ┆ 80001 `
` pledgeCollateral                           ┆ 62939           ┆ 65049  ┆ 65118  ┆ 509232  ┆ 80001  `

`removeAllQuoteToken                        ┆ 149419          ┆ 151591 ┆ 151296 ┆ 168369  ┆ 2000`
`removeAllQuoteToken                        ┆ 149633          ┆ 151774 ┆ 151468 ┆ 168951  ┆ 2000`

`removeQuoteToken                           ┆ 184196          ┆ 190028 ┆ 189821 ┆ 209409  ┆ 2000` 
`removeQuoteToken                           ┆ 186410          ┆ 192242 ┆ 192035 ┆ 212019  ┆ 2000`

`repay                                      ┆ 483298          ┆ 558034 ┆ 569953 ┆ 809532  ┆ 16001`
`repay                                      ┆ 483511          ┆ 558247 ┆ 570166 ┆ 809745  ┆ 16001 `

`take                                       ┆ 50856           ┆ 54687  ┆ 54169  ┆ 100252  ┆ 15998`
`take                                       ┆ 51251           ┆ 55003  ┆ 54485  ┆ 100568  ┆ 15998 `

`kick                                       ┆ 173475          ┆ 249007 ┆ 244636 ┆ 1078440 ┆ 16000`
`kick                                       ┆ 173828          ┆ 249362 ┆ 244989 ┆ 1100693 ┆ 16000   │`

Contract size:
```
============ Deployment Bytecode Sizes ============
  ERC721Pool               -  24,423B  (99.37%)
  ERC20Pool                -  24,040B  (97.82%)
  PositionManager          -  15,560B  (63.31%)
  PoolInfoUtils            -   7,049B  (28.68%)
  BucketMath               -   6,085B  (24.76%)
  ERC721                   -   4,421B  (17.99%)
  ERC721PoolFactory        -   2,378B  (9.68%)
  ERC20                    -   2,140B  (8.71%)
  ERC20PoolFactory         -   1,831B  (7.45%)
  Deposits                 -     173B  (0.70%)
  Address                  -      86B  (0.35%)
  Auctions                 -      86B  (0.35%)
  Base64                   -      86B  (0.35%)
  Buckets                  -      86B  (0.35%)
  ClonesWithImmutableArgs  -      86B  (0.35%)
  EnumerableSet            -      86B  (0.35%)
  Loans                    -      86B  (0.35%)
  Maths                    -      86B  (0.35%)
  PRBMath                  -      86B  (0.35%)
  PRBMathSD59x18           -      86B  (0.35%)
  PRBMathUD60x18           -      86B  (0.35%)
  PoolUtils                -      86B  (0.35%)
  SafeERC20                -      86B  (0.35%)
  Strings                  -      86B  (0.35%)
  Clone                    -      63B  (0.26%)

```

NOTE: Attempted to store the `topOfBookDownAuctionDebtIndex` as a way to save gas and reduce the total number of `deposits.findIndexOfSum` calls... realized that implementing this way stops the `_isAuctionDebtLocked` check from including interest when it finds the index, thus not matching expected functionality.

### CODE REVIEW

File: src/base/interfaces/pool/IPoolErrors.sol [hunk 1]
```diff
@@ -113,4 +113,10 @@ interface IPoolErrors {
      *  @notice Borrower is attempting to borrow an amount of quote tokens that will push the pool into under-collateralization.
      */
     error PoolUnderCollateralized();
+
+
+    /**
+     *  @notice Lender is attempting to remove quote tokens from a bucket that exists above active auction debt from top-of-book downward.
+     */
+    error DepositLockedByAuctionDebt();
```
Comment (EdNoepel) [inline src/base/interfaces/pool/IPoolErrors.sol]: Noun/verb confusion in the naming.  We're blocking the ability to _withdraw_, yet error could be interpreted as a user cannot _deposit_ into the bucket.  That said, I wouldn't want _withdrawal_ in the name, because our contracts call it _remove_.  `RemovalOfDepositLockedByAuctionDebt` is a bit wordy.  Thoughts?

File: src/base/Pool.sol [hunk 2]
```diff
@@ -973,4 +989,18 @@ abstract contract Pool is Clone, Multicall, IPool {
...
+     *  @notice Called by LPB removal functions assess whether or not LPB is locked.
+     *  @param  index_   The bucket index from which LPB is attempting to be removed.
+     *  @param  inflator_ The pool inflator used to properly assess t0DebtInAuction.
+     */
+    function _isAuctionDebtLocked(
+        uint256 index_,
+        uint256 inflator_
+    ) internal view returns (bool) {
+        // deposit in buckets within liquidation debt from the top-of-book down are frozen.
+        return index_ <= deposits.findIndexOfSum(Maths.wmul(t0DebtInAuction, inflator_));
```
Comment (EdNoepel) [inline src/base/Pool.sol]: `findIndexOfSum` is an expensive operation, and we run this every time someone tries to remove quote token from a bucket.  Would it make sense to add a conditional above which could cheaply return `false` if `t0DebtInAuction==0`?

File: src/base/Pool.sol [hunk 3]
```diff
@@ -457,14 +464,19 @@ abstract contract Pool is Clone, Multicall, IPool {
 
         uint256 newLup = _lup(poolState.accruedDebt);
 
+        uint256 collateralization = _collateralization(
+                    Maths.wmul(borrower.t0debt, poolState.inflator),
+                    borrower.collateral,
+                    newLup
+                );
+
+        if (collateralization >= Maths.WAD && auctions.isActive(borrowerAddress_)) t0DebtInAuction -= borrower.t0debt;
```
Comment (EdNoepel) [inline src/base/Pool.sol]: I needed to think about this, but it makes sense.  Borrower can add collateral while a liquidation is going, but it won't affect the amount of debt in the auction until they add enough to become overcollateralized.  Opposite is true in `_payLoan`; every repayment decreases the auction debt.

We could have `auctions.checkAndRemove` maintain this, but it would waste gas on unwanted storage writes.  I'm happy with it as-is.

**Review by EdNoepel — COMMENTED** [review]
Looks good, but I added a few comments for discussion.

File: src/base/interfaces/pool/IPoolErrors.sol (hunk 1 above)
Comment (ith-harvey) [reply src/base/interfaces/pool/IPoolErrors.sol]: Agreed, renamed to `RemoveDepositLockedByAuctionDebt` here -> https://github.com/ajna-finance/contracts/pull/310/commits/237bbd714be4af424424ad3bed6d63617a1b0d32

File: src/base/Pool.sol (hunk 2 above)
Comment (ith-harvey) [reply src/base/Pool.sol]: Agree, updated here -> https://github.com/ajna-finance/contracts/pull/310/commits/237bbd714be4af424424ad3bed6d63617a1b0d32

File: src/base/Pool.sol (hunk 3 above)
Comment (ith-harvey) [reply src/base/Pool.sol]: :+1:

**Review by EdNoepel — COMMENTED** [review]
Looks good; have you updated gas numbers with the most recent commit?

**Comment by ith-harvey** [comment]
> Looks good; have you updated gas numbers with the most recent commit?

updated gas numbers :point_up:  : )

Before Eds change:
```
│ kick                                          ┆ 173828          ┆ 249362 ┆ 244989 ┆ 1100693 ┆ 16000   │
│ moveQuoteToken                    ┆ 286872          ┆ 286872 ┆ 286872 ┆ 286872  ┆ 1       │
│ pledgeCollateral                      ┆ 62939           ┆ 65049  ┆ 65118  ┆ 509232  ┆ 80001   │
│ removeAllQuoteToken             ┆ 159707          ┆ 161730 ┆ 161588 ┆ 180513  ┆ 2000    │
│ removeQuoteToken                 ┆ 206062          ┆ 235298 ┆ 235500 ┆ 242256  ┆ 2000    │
│ repay                                      ┆ 483514          ┆ 558250 ┆ 570169 ┆ 809748  ┆ 16001   │
│ take                                       ┆ 50746           ┆ 54599  ┆ 54081  ┆ 100164  ┆ 15998   │
```
VS 
After Eds suggested change (current):
```
│ kick                                        ┆ 173828          ┆ 249362 ┆ 244989 ┆ 1100693 ┆ 16000   │
│ moveQuoteToken                  ┆ 279310          ┆ 279310 ┆ 279310 ┆ 279310  ┆ 1       │
│ pledgeCollateral                    ┆ 62939           ┆ 65049  ┆ 65118  ┆ 509232  ┆ 80001   │
│ removeAllQuoteToken           ┆ 149633          ┆ 151774 ┆ 151468 ┆ 168951  ┆ 2000    │
│ removeQuoteToken               ┆ 186410          ┆ 192242 ┆ 192035 ┆ 212019  ┆ 2000    │
│ repay                                     ┆ 483511          ┆ 558247 ┆ 570166 ┆ 809745  ┆ 16001   │
│ take                                       ┆ 51251           ┆ 55003  ┆ 54485  ┆ 100568  ┆ 15998   │
```

Appears `take` increased which is interesting...

**Review by EdNoepel — APPROVED** [review]
Gas numbers look much better.

**Comment by ith-harvey** [comment]
Georges Private chat feedback:
* Implement `_checkIfAuctionDebtLock` ->

**Comment by ith-harvey** [comment]
remove `debt()` method change -> https://github.com/ajna-finance/contracts/pull/310/commits/a24402da127e1f90f8ad78bc5a88f588fa48c955

new Gas results:
│ kick                                       ┆ 173828          ┆ 249362 ┆ 244989 ┆ 1100693 ┆ 16000   │
│ loansInfo                                  ┆ 2684            ┆ 2685   ┆ 2684   ┆ 10684   ┆ 48013   │
│ moveQuoteToken                             ┆ 279281          ┆ 279281 ┆ 279281 ┆ 279281  ┆ 1       │
│ pledgeCollateral                           ┆ 62939           ┆ 65049  ┆ 65118  ┆ 509232  ┆ 80001   │
│ removeAllQuoteToken                        ┆ 149604          ┆ 151749 ┆ 151444 ┆ 168922  ┆ 2000    │
│ removeQuoteToken                           ┆ 186381          ┆ 192213 ┆ 192006 ┆ 211990  ┆ 2000    │
│ repay                                      ┆ 483489          ┆ 558225 ┆ 570144 ┆ 809723  ┆ 16001   │
│ take                                       ┆ 51251           ┆ 55003  ┆ 54485  ┆ 100568  ┆ 15998   │

contract size results:
```
============ Deployment Bytecode Sizes ============
  ERC721Pool               -  24,235B  (98.61%)
  ERC20Pool                -  23,852B  (97.05%)
  PositionManager          -  15,492B  (63.03%)
  PoolInfoUtils            -   7,153B  (29.10%)
  BucketMath               -   6,085B  (24.76%)
  ERC721                   -   4,421B  (17.99%)
  ERC721PoolFactory        -   2,378B  (9.68%)
  ERC20                    -   2,140B  (8.71%)
  ERC20PoolFactory         -   1,831B  (7.45%)
```

File: src/base/Pool.sol [hunk 4]
```diff
@@ -973,4 +984,20 @@ abstract contract Pool is Clone, Multicall, IPool {
     function quoteTokenScale() external pure override returns (uint256) {
         return _getArgUint256(40);
     }
+
+    /**
+     *  @notice Called by LPB removal functions assess whether or not LPB is locked.
+     *  @param  index_   The bucket index from which LPB is attempting to be removed.
+     *  @param  inflator_ The pool inflator used to properly assess t0DebtInAuction.
+     */
+    function _checkIfAuctionDebtLocked(
```
Comment (grandizzy) [inline src/base/Pool.sol]: maybe rename to `_revertOnLockedAuctionDebt` or smth to contain revert in name?

File: src/_test/ERC20Pool/ERC20PoolPrecision.t.sol [hunk 5]
```diff
@@ -473,8 +475,9 @@ contract ERC20PoolPrecisionTest is ERC20DSTestPlus {
                 maxThresholdPrice: 3_025.946482308870941594 * 1e18
             }
         );
+        (poolDebt,,) = _pool.debtInfo();
         assertEq(_pool.depositSize(),       150_000 * 1e18);
-        assertEq(_pool.debt(),              debt);
+        assertEq(poolDebt,              debt);
```
Comment (grandizzy) [inline src/_test/ERC20Pool/ERC20PoolPrecision.t.sol]: NIT alignment

File: src/_test/ERC20Pool/ERC20PoolPrecision.t.sol [hunk 6]
```diff
@@ -407,8 +408,9 @@ contract ERC20PoolPrecisionTest is ERC20DSTestPlus {
                 maxThresholdPrice: 100.192307692307692400 * 1e18
             }
         );
+        (poolDebt,,) = _pool.debtInfo();
         assertEq(_pool.depositSize(),       150_000 * 1e18);
-        assertEq(_pool.debt(),              debt);
+        assertEq(poolDebt,              debt);
```
Comment (grandizzy) [inline src/_test/ERC20Pool/ERC20PoolPrecision.t.sol]: NIT alignment

File: src/_test/ERC20Pool/ERC20PoolPrecision.t.sol [hunk 7]
```diff
@@ -338,8 +338,9 @@ contract ERC20PoolPrecisionTest is ERC20DSTestPlus {
                 maxThresholdPrice: 200.192307692307692400 * 1e18
             }
         );
+        (uint256 poolDebt,,) = _pool.debtInfo();
         assertEq(_pool.depositSize(),       150_000 * _quotePoolPrecision);
-        assertEq(_pool.debt(),              debt);
+        assertEq(poolDebt,              debt);
```
Comment (grandizzy) [inline src/_test/ERC20Pool/ERC20PoolPrecision.t.sol]: NIT alignment

File: src/libraries/Loans.sol [hunk 8]
```diff
@@ -68,14 +68,15 @@ library Loans {
...
-    ) internal returns (uint256 kickPenalty_) {
+    ) internal returns (uint256 kickPenalty_, uint256 totalT0BorrowerDebt_) {
         // update loan heap
         _remove(self, borrower_);
 
         // update borrower balance
         Borrower storage borrower = self.borrowers[borrower_];
         kickPenalty_              = Maths.wmul(Maths.wdiv(rate_, 4 * 1e18), debt_); // when loan is kicked, penalty of three months of interest is added
-        borrower.t0debt           = Maths.wdiv(debt_ + kickPenalty_, inflator_);
+        totalT0BorrowerDebt_      = Maths.wdiv(debt_ + kickPenalty_, inflator_);
```
Comment (grandizzy) [inline src/libraries/Loans.sol]: we shouldn't return `totalT0BorrowerDebt_` here, but leave `internal returns (uint256 kickPenalty_)` and calculate / return `kickPenalty` at t0
```
kickPenalty_       = Maths.wdiv(Maths.wmul(Maths.wdiv(rate_, 4 * 1e18), debt_), inflator_);
borrower.t0debt  += kickPenalty_;

File: src/base/Pool.sol [hunk 9]
```diff
@@ -371,14 +378,14 @@ abstract contract Pool is Clone, Multicall, IPool {
             ) >= Maths.WAD
         ) revert BorrowerOk();
 
-        uint256 kickPenalty = loans.kick(
+        (uint256 kickPenalty, uint256 t0borrowerDebt)= loans.kick(
```
Comment (grandizzy) [inline src/base/Pool.sol]: this should return only one variable `uint256 kickPenaltyT0 = loans.kick()` (see comment I left in `loans.kick`)

File: src/base/Pool.sol [hunk 10]
```diff
@@ -371,14 +378,14 @@ abstract contract Pool is Clone, Multicall, IPool {
...
             borrowerAddress_,
             borrowerDebt,
             poolState.inflator,
             poolState.rate
         );
         poolState.accruedDebt += kickPenalty;
```
Comment (grandizzy) [inline src/base/Pool.sol]: `poolState.accruedDebt += kickPenalty;` should be OK as long as `kickPenalty` is actually the `kickPenaltyT0`

File: src/base/Pool.sol [hunk 11]
```diff
@@ -371,14 +378,14 @@ abstract contract Pool is Clone, Multicall, IPool {
...
-        t0poolDebt += Maths.wdiv(kickPenalty, poolState.inflator);
+        t0poolDebt            += Maths.wdiv(kickPenalty, poolState.inflator);
```
Comment (grandizzy) [inline src/base/Pool.sol]: this should be `t0poolDebt            += kickPenalty`

File: src/base/Pool.sol [hunk 12]
```diff
@@ -389,6 +396,7 @@ abstract contract Pool is Clone, Multicall, IPool {
         );
 
         // update pool state
+        t0DebtInAuction += t0borrowerDebt;
```
Comment (grandizzy) [inline src/base/Pool.sol]: this should be `t0DebtInAuction += borrower.t0debt + kickPenalty;` where `kickPenalty` is actually `kickPenaltyT0` returned by kick

**Review by grandizzy — CHANGES_REQUESTED** [review]
I think `Loans.kick` should not return `internal returns (uint256 kickPenalty_, uint256 totalT0BorrowerDebt_) ` but rather only `kickPenalty` calculated at t0

File: src/base/Pool.sol (hunk 4 above)
Comment (ith-harvey) [reply src/base/Pool.sol]: How about `_revertIfAuctionDebtLocked`?

File: src/libraries/Loans.sol (hunk 8 above)
Comment (ith-harvey) [reply src/libraries/Loans.sol]: Since borrower is read from storage on LN 76, I think this proposal should work.

File: src/base/Pool.sol (hunk 10 above)
Comment (ith-harvey) [reply src/base/Pool.sol]: I don't think this is correct... `poolState.accruedDebt` is not t0'd, see -> https://github.com/ajna-finance/contracts/blob/544a2cadff454d1ff7c0c53b378258e284aa0da9/src/base/Pool.sol#L632

For this reason I think we should actually do `poolState.accruedDebt += Maths.wmul(kickPenalty, poolInflator);`

File: src/base/Pool.sol (hunk 12 above)
Comment (ith-harvey) [reply src/base/Pool.sol]: How do you propose retrieving `borrower.t0debt`? Up above its retreived from memory meaning we would need to re-retreive the `borrower` struct... is that what you are proposing?

Comment (grandizzy) [reply src/base/Pool.sol]: we could remove `Loan.kick` altogether and calculate penalty, update borrower debt and call `loan.remove` from `Pool.kick` function.

**Comment by ith-harvey** [comment]
Per George's suggested revisions -> https://github.com/ajna-finance/contracts/pull/310/commits/3c9d81703c240fac6add52aeff18e5afbbfe13e5

New Gas results:
`│ kick                                       ┆ 173746          ┆ 248751 ┆ 244377 ┆ 1098894 ┆ 16000   │`
`│ loansInfo                                  ┆ 2684            ┆ 2685   ┆ 2684   ┆ 10684   ┆ 48013   │`
`│ moveQuoteToken                             ┆ 287583          ┆ 287583 ┆ 287583 ┆ 287583  ┆ 1       │`
`│ pledgeCollateral                           ┆ 62434           ┆ 64503  ┆ 64613  ┆ 540564  ┆ 80001   │`
`│ removeAllQuoteToken                        ┆ 160174          ┆ 165574 ┆ 165632 ┆ 171202  ┆ 2000    │`
`│ removeQuoteToken                           ┆ 185229          ┆ 191061 ┆ 190854 ┆ 210248  ┆ 2000    │`
`│ repay                                      ┆ 514971          ┆ 589528 ┆ 601387 ┆ 840966  ┆ 16001   │`
`│ take                                       ┆ 51277           ┆ 55125  ┆ 54506  ┆ 274692  ┆ 15998   │`


Contract size reduction:
```
============ Deployment Bytecode Sizes ============
  ERC721Pool               -  24,221B  (98.55%)
  ERC20Pool                -  23,833B  (96.97%)
  PositionManager          -  15,492B  (63.03%)
  PoolInfoUtils            -   7,153B  (29.10%)
  BucketMath               -   6,085B  (24.76%)
  ERC721                   -   4,421B  (17.99%)
  ERC721PoolFactory        -   2,378B  (9.68%)
  ERC20                    -   2,140B  (8.71%)
  ERC20PoolFactory         -   1,831B  (7.45%)
```

File: src/base/Pool.sol [hunk 13]
```diff
@@ -395,8 +396,18 @@ abstract contract Pool is Clone, Multicall, IPool {
             deposits.momp(poolState.accruedDebt, loans.noOfLoans())
         );
 
+        // update borrower & pool debt with kickPenalty
+        uint256 kickPenalty    =  Maths.wmul(Maths.wdiv(poolState.rate, 4 * 1e18), borrowerDebt); // when loan is kicked, penalty of three months of interest is added 
+        borrowerDebt           += kickPenalty;
+        poolState.accruedDebt  += kickPenalty;
+        t0poolDebt             += Maths.wdiv(kickPenalty, poolState.inflator);
+
+        uint256 t0BorrowerDebt =  Maths.wdiv(borrowerDebt, poolState.inflator);
```
Comment (EdNoepel) [inline src/base/Pool.sol]: This is what Matt wants us to avoid.  Instead of calculating in current terms and overwriting, should calculate in t0 terms and then adjust pool and borrower by same amount.

```
        uint256 kickPenalty    =  Maths.wmul(Maths.wdiv(poolState.rate, 4 * 1e18), borrowerDebt);
        borrowerDebt           += kickPenalty;
        poolState.accruedDebt  += kickPenalty; 

        uint256 t0kickPenalty                    =  Maths.wdiv(kickPenalty, poolState.inflator);
        t0poolDebt                               += t0kickPenalty;
        // TODO: avoid double storage hit
        loans.borrowers[borrowerAddress_].t0debt += t0kickPenalty;     
        t0DebtInAuction                          += loans.borrowers[borrowerAddress_].t0debt;
```

**Review by EdNoepel — CHANGES_REQUESTED** [review]

File: src/base/Pool.sol (hunk 12 above)
Comment (ith-harvey) [reply src/base/Pool.sol]: Agree -> https://github.com/ajna-finance/contracts/commit/3c9d81703c240fac6add52aeff18e5afbbfe13e5

File: src/base/Pool.sol (hunk 11 above)
Comment (ith-harvey) [reply src/base/Pool.sol]: agree -> https://github.com/ajna-finance/contracts/commit/3c9d81703c240fac6add52aeff18e5afbbfe13e5

File: src/base/Pool.sol (hunk 10 above)
Comment (ith-harvey) [reply src/base/Pool.sol]: done -> https://github.com/ajna-finance/contracts/commit/3c9d81703c240fac6add52aeff18e5afbbfe13e5

File: src/base/Pool.sol (hunk 9 above)
Comment (ith-harvey) [reply src/base/Pool.sol]: included here -> https://github.com/ajna-finance/contracts/commit/3c9d81703c240fac6add52aeff18e5afbbfe13e5

File: src/_test/ERC20Pool/ERC20PoolPrecision.t.sol (hunk 7 above)
Comment (ith-harvey) [reply src/_test/ERC20Pool/ERC20PoolPrecision.t.sol]: cleaned up -> https://github.com/ajna-finance/contracts/commit/3c9d81703c240fac6add52aeff18e5afbbfe13e5

File: src/_test/ERC20Pool/ERC20PoolPrecision.t.sol (hunk 6 above)
Comment (ith-harvey) [reply src/_test/ERC20Pool/ERC20PoolPrecision.t.sol]: cleaned up -> https://github.com/ajna-finance/contracts/commit/3c9d81703c240fac6add52aeff18e5afbbfe13e5

File: src/base/Pool.sol (hunk 13 above)
Comment (ith-harvey) [reply src/base/Pool.sol]: Revised -> https://github.com/ajna-finance/contracts/pull/310/commits/97ff82d27196d4da6a6ce17e7a5da3b642f76d5d

**Comment by ith-harvey** [comment]
Gas with Eds suggested change -> https://github.com/ajna-finance/contracts/pull/310/commits/97ff82d27196d4da6a6ce17e7a5da3b642f76d5d
`kick                                       ┆ 173491          ┆ 252090 ┆ 246789 ┆ 1098639 ┆ 32000   `

Contract size exceeding is not due to the code changes but rather to merging after George's PR (https://github.com/ajna-finance/contracts/commit/cca431280cc21a76719d4804fabc1e5101337669) :
```
  ERC20Pool                -  24,764B  (100.76%)
  ERC721Pool               -  24,295B  (98.85%)
  PositionManager          -  15,492B  (63.03%)
  PoolInfoUtils            -   7,153B  (29.10%)
  BucketMath               -   6,085B  (24.76%)
  ERC721                   -   4,421B  (17.99%)
  ERC721PoolFactory        -   2,378B  (9.68%)
  ERC20                    -   2,140B  (8.71%)
  ERC20PoolFactory         -   1,831B  (7.45%)
```

Contract reduction notes:
- attempted to remove `actions.isActive()` -> resulted in no reduction.

**Review by EdNoepel — APPROVED** [review]
👌

File: src/libraries/Auctions.sol [hunk 14]
```diff
@@ -457,7 +457,20 @@ library Auctions {
...
+        if (isActive(self, borrower_)) revert AuctionActive();
+    }
+
+    /**
+     *  @notice Returns true if borrower is in auction.
+     *  @dev    Used to accuratley increment and decrement t0DebtInAuction.
+     *  @param  borrower_ Borrower address to check auction status for.
+     *  @return  active_ Boolean, based on if borrower is in auction.
+     */
+    function isActive(
```
Comment (MikeHathaway) [inline src/libraries/Auctions.sol]: NIT: Can update the internal helper functions with the underscore prefix, so `_isActive`

File: src/_test/ERC721Pool/ERC721PoolLiquidations.t.sol [hunk 15]
```diff
@@ -187,15 +188,15 @@ contract ERC721PoolLiquidationsTest is ERC721HelperContract {
...
                 bond:       0.227287198298417188 * 1e18
             }
         );
 
-        /******************************/
-        /*** Assert Post-kick state ***/
-        /******************************/
+        // /******************************/
+        // /*** Assert Post-kick state ***/
+        // /******************************/
```
Comment (MikeHathaway) [inline src/_test/ERC721Pool/ERC721PoolLiquidations.t.sol]: NIT: was this style change intentional?

**Review by MikeHathaway — APPROVED** [review]
LGTM! Couple minor stylistic questions, but didn't see anything else

File: src/_test/ERC20Pool/ERC20PoolPrecision.t.sol (hunk 5 above)
Comment (ith-harvey) [reply src/_test/ERC20Pool/ERC20PoolPrecision.t.sol]: cleaned up -> https://github.com/ajna-finance/contracts/commit/3c9d81703c240fac6add52aeff18e5afbbfe13e5

**Review by grandizzy — APPROVED** [review]
